composable_kernel/include/tensor_operation/reduction_functions_blockwise.hpp

```diff
@@ -23,247 +23,145 @@
+#include "reduction_functions_accumulate.hpp"
 
 namespace ck {
 
-template <typename buffer2dDescType,
-          bool blockIsOneRow,
+template <typename buffer1dDescType,
+          index_t BlockSize,
+          index_t dim0_thread_cluster_length,
+          index_t dim1_thread_cluster_length,
+          bool reorder_thread_clusters,
           typename opReduce,
-          NanPropagation_t nanPropaOpt>
-struct BlockwiseReduction_2d_block_buffer
+          bool propagate_nan>
+struct PartitionedBlockwiseReduction_1d_block_buffer
 {
     using compType = typename opReduce::dataType;
 
-    static constexpr auto buffer2dDesc = buffer2dDescType{};
+    static constexpr auto buffer1dDesc = buffer1dDescType{};
 
-    static constexpr index_t BlockSize =
-        blockIsOneRow ? buffer2dDesc.GetLength(Number<1>{}) : buffer2dDesc.GetLength(Number<0>{});
-    static constexpr index_t NumBlocks =
-        blockIsOneRow ? buffer2dDesc.GetLength(Number<0>{}) : buffer2dDesc.GetLength(Number<1>{});
-    using binop = detail::binop_with_nan_check<nanPropaOpt, opReduce, compType>;
+    static_assert(BlockSize == dim0_thread_cluster_length * dim1_thread_cluster_length,
+                  "The product of cluster lengths should be same as BlockSize!");
+    static_assert(dim1_thread_cluster_length > 1,
+                  "Parallel reduction need work on at least two elements");
 
-    // This interface does not accumulate on indices
-    template <typename BufferType>
-    __device__ static void
-    Reduce(BufferType& block_buffer, index_t toReduceBlocks, compType& accuData)
-    {
-        const index_t thread_local_id = get_thread_local_1d_id();
-        compType lAccuData            = opReduce::GetReductionZeroVal();
-
-        index_t offset;
-        for(index_t otherDimInd = 0; otherDimInd < toReduceBlocks; otherDimInd++)
-        {
-            offset = blockIsOneRow
-                         ? buffer2dDesc.CalculateOffset(make_tuple(otherDimInd, thread_local_id))
-                         : buffer2dDesc.CalculateOffset(make_tuple(thread_local_id, otherDimInd));
-            compType opData = type_convert<compType>(block_buffer[offset]);
-
-            binop::calculate(lAccuData, opData);
-        }
+    static_assert(buffer1dDesc.GetElementSize() == BlockSize,
+                  "The buffer size should be the same as BlockSize!");
 
-        offset = blockIsOneRow ? buffer2dDesc.CalculateOffset(make_tuple(0, thread_local_id))
-                               : buffer2dDesc.CalculateOffset(make_tuple(thread_local_id, 0));
+    using Accumulation = detail::accumulate_with_nan_check<propagate_nan, opReduce, compType>;
 
-        block_buffer(offset) = lAccuData;
+    template <typename BufferType>
+    __device__ static void Reduce(BufferType& block_buffer,
+                                  compType& accuData,
+                                  index_t thread_dim0_cluster_id,
+                                  index_t thread_dim1_cluster_id)
```

Comment:
thread cluster id should be calculated inside this class, by using get_thread_local_1d_id and cluster descriptor

https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/22d438ae9e02b674e1656263df07799ee06fb466/composable_kernel/include/tensor_operation/blockwise_tensor_slice_transfer_v4r1.hpp#L78-L79

Reply:
I see